File: conformance/runners/godj/scenarios.go
package godj

import (
	"context"
	"errors"
	"fmt"

	"github.com/progresshans/godj/conformance/internal/protocol"
	"github.com/progresshans/godj/db/sqlite"
	"github.com/progresshans/godj/examples/article/models"
	"github.com/progresshans/godj/orm"
	"github.com/progresshans/godj/query"
)

type databaseScenario func(context.Context, *sqlite.Backend) (protocol.Observation, error)

func withArticleDatabase(ctx context.Context, contractID string, scenario databaseScenario) (protocol.Observation, error) {
	backend, err := openArticleDatabase(ctx, contractID)
	if err != nil {
		return protocol.Observation{}, err
	}
	observation, scenarioErr := scenario(ctx, backend)
	closeErr := backend.Close()
	if scenarioErr != nil {
		return protocol.Observation{}, errors.Join(scenarioErr, closeErr)
	}
	if closeErr != nil {
		return protocol.Observation{}, closeErr
	}
	return observation, nil
}

func queryExact(ctx context.Context, contractID string) (protocol.Observation, error) {
	return withArticleDatabase(ctx, contractID, func(ctx context.Context, backend *sqlite.Backend) (protocol.Observation, error) {
		articles, err := models.ArticleObjects.Using(backend).
			Filter(models.ArticleFields.Title.Exact("Alpine Guide")).
			OrderBy(models.ArticleFields.ID.Asc()).
			All(ctx)
		if err != nil {
			return protocol.Observation{}, err
		}
		return resultObservation(ctx, backend, contractID, protocol.PhaseEvaluation, articleList(articles), nil)
	})
}

func queryASCIIInsensitiveContains(ctx context.Context, contractID string) (protocol.Observation, error) {
	return withArticleDatabase(ctx, contractID, func(ctx context.Context, backend *sqlite.Backend) (protocol.Observation, error) {
		articles, err := models.ArticleObjects.Using(backend).
			Filter(models.ArticleFields.Title.IContains("django")).
			OrderBy(models.ArticleFields.ID.Asc()).
			All(ctx)
		if err != nil {
			return protocol.Observation{}, err
		}
		return resultObservation(ctx, backend, contractID, protocol.PhaseEvaluation, articleList(articles), nil)
	})
}

func queryChainedAnd(ctx context.Context, contractID string) (protocol.Observation, error) {
	return withArticleDatabase(ctx, contractID, func(ctx context.Context, backend *sqlite.Backend) (protocol.Observation, error) {
		articles, err := models.ArticleObjects.Using(backend).
			Filter(models.ArticleFields.Title.IContains("django")).
			Filter(models.ArticleFields.Published.Exact(true)).
			OrderBy(models.ArticleFields.ID.Asc()).
			All(ctx)
		if err != nil {
			return protocol.Observation{}, err
		}
		return resultObservation(ctx, backend, contractID, protocol.PhaseEvaluation, articleList(articles), nil)
	})
}

func queryChainPreservesSource(ctx context.Context, contractID string) (protocol.Observation, error) {
	return withArticleDatabase(ctx, contractID, func(ctx context.Context, backend *sqlite.Backend) (protocol.Observation, error) {
		base := models.ArticleObjects.Using(backend).
			Filter(models.ArticleFields.Published.Exact(true))
		derived := base.Filter(models.ArticleFields.Title.Exact("Django Deep Dive"))

		baseArticles, err := base.OrderBy(models.ArticleFields.ID.Asc()).All(ctx)
		if err != nil {
			return protocol.Observation{}, fmt.Errorf("evaluate base query: %w", err)
		}
		derivedArticles, err := derived.OrderBy(models.ArticleFields.ID.Asc()).All(ctx)
		if err != nil {
			return protocol.Observation{}, fmt.Errorf("evaluate derived query: %w", err)
		}
		result := protocol.Object(map[string]protocol.Value{
			"base":    articleList(baseArticles),
			"derived": articleList(derivedArticles),
		})
		return resultObservation(ctx, backend, contractID, protocol.PhaseEvaluation, result, nil)
	})
}

func queryOrderAndLimit(ctx context.Context, contractID string) (protocol.Observation, error) {
	return withArticleDatabase(ctx, contractID, func(ctx context.Context, backend *sqlite.Backend) (protocol.Observation, error) {
		querySet := models.ArticleObjects.Using(backend).
			OrderBy(models.ArticleFields.Title.Desc(), models.ArticleFields.ID.Asc())
		querySet, err := querySet.Limit(2)
		if err != nil {
			return protocol.Observation{}, err
		}
		articles, err := querySet.All(ctx)
		if err != nil {
			return protocol.Observation{}, err
		}
		return resultObservation(ctx, backend, contractID, protocol.PhaseEvaluation, articleList(articles), nil)
	})
}

func queryEmptyResult(ctx context.Context, contractID string) (protocol.Observation, error) {
	return withArticleDatabase(ctx, contractID, func(ctx context.Context, backend *sqlite.Backend) (protocol.Observation, error) {
		articles, err := models.ArticleObjects.Using(backend).
			Filter(models.ArticleFields.Title.Exact("missing")).
			OrderBy(models.ArticleFields.ID.Asc()).
			All(ctx)
		if err != nil {
			return protocol.Observation{}, err
		}
		return resultObservation(ctx, backend, contractID, protocol.PhaseEvaluation, articleList(articles), nil)
	})
}

func queryIsNull(ctx context.Context, contractID string) (protocol.Observation, error) {
	return withArticleDatabase(ctx, contractID, func(ctx context.Context, backend *sqlite.Backend) (protocol.Observation, error) {
		base := models.ArticleObjects.Using(backend)
		nonNull, err := base.
			Filter(models.ArticleFields.Summary.IsNull(false)).
			OrderBy(models.ArticleFields.ID.Asc()).
			All(ctx)
		if err != nil {
			return protocol.Observation{}, fmt.Errorf("evaluate isnull=false query: %w", err)
		}
		null, err := base.
			Filter(models.ArticleFields.Summary.IsNull(true)).
			OrderBy(models.ArticleFields.ID.Asc()).
			All(ctx)
		if err != nil {
			return protocol.Observation{}, fmt.Errorf("evaluate isnull=true query: %w", err)
		}
		result := protocol.Object(map[string]protocol.Value{
			"false": articleList(nonNull),
			"true":  articleList(null),
		})
		return resultObservation(ctx, backend, contractID, protocol.PhaseEvaluation, result, nil)
	})
}

func queryUnknownField(ctx context.Context, contractID string) (protocol.Observation, error) {
	return dynamicErrorObservation(ctx, contractID, orm.LookupInput{Key: "unknown_field", Value: "value"}, query.CodeUnknownField)
}

func queryUnsupportedLookup(ctx context.Context, contractID string) (protocol.Observation, error) {
	return dynamicErrorObservation(ctx, contractID, orm.LookupInput{Key: "title__starts", Value: "Django"}, query.CodeUnsupportedLookup)
}

func dynamicErrorObservation(ctx context.Context, contractID string, input orm.LookupInput, expectedCode string) (protocol.Observation, error) {
	return withArticleDatabase(ctx, contractID, func(ctx context.Context, backend *sqlite.Backend) (protocol.Observation, error) {
		before := backend.QueryCount()
		_, parseErr := orm.ParseDynamic(models.ArticleDescriptor{}, nil, []orm.LookupInput{input})
		after := backend.QueryCount()
		if after != before {
			return protocol.Observation{}, fmt.Errorf("dynamic lookup construction issued %d query calls", after-before)
		}
		var queryErr *query.Error
		if !errors.As(parseErr, &queryErr) {
			return protocol.Observation{}, fmt.Errorf("dynamic lookup error = %v, want *query.Error", parseErr)
		}
		if queryErr.Code != expectedCode {
			return protocol.Observation{}, fmt.Errorf("dynamic lookup error code = %q, want %q", queryErr.Code, expectedCode)
		}
		state, err := readDatabaseState(ctx, backend)
		if err != nil {
			return protocol.Observation{}, err
		}
		return protocol.Observation{
			ID:     contractID,
			Status: protocol.StatusObserved,
			Phase:  protocol.PhaseConstruction,
			Error: &protocol.ObservedError{
				Category:          queryErr.Category,
				Code:              queryErr.Code,
				Message:           queryErr.Error(),
				MessageIsContract: boolPointer(false),
			},
			DBState: valuePointer(state),
		}, nil
	})
}

func queryConstructionHasNoIO(ctx context.Context, contractID string) (protocol.Observation, error) {
	return withArticleDatabase(ctx, contractID, func(ctx context.Context, backend *sqlite.Backend) (protocol.Observation, error) {
		before := backend.QueryCount()
		querySet := models.ArticleObjects.Using(backend).
			Filter(models.ArticleFields.Published.Exact(true)).
			OrderBy(models.ArticleFields.ID.Asc())
		querySet, err := querySet.Limit(2)
		if err != nil {
			return protocol.Observation{}, err
		}
		queriesDuringConstruction := backend.QueryCount() - before
		if queriesDuringConstruction != 0 {
			return protocol.Observation{}, fmt.Errorf("query construction issued %d query calls", queriesDuringConstruction)
		}
		// Keep the value live so this observation verifies construction rather
		// than a compiler-elided expression whose result is discarded.
		constructed := querySet.Plan().Table() != ""
		result := protocol.Object(map[string]protocol.Value{
			"queryset_constructed": protocol.Boolean(constructed),
		})
		metrics := protocol.Object(map[string]protocol.Value{
			"queries_during_construction": protocol.Integer(fmt.Sprint(queriesDuringConstruction)),
		})
		return resultObservation(ctx, backend, contractID, protocol.PhaseConstruction, result, &metrics)
	})
}

func schemaModelMetadata(contractID string) (protocol.Observation, error) {
	result, err := metadataValue((models.ArticleDescriptor{}).Metadata())
	if err != nil {
		return protocol.Observation{}, err
	}
	return protocol.Observation{
		ID:     contractID,
		Status: protocol.StatusObserved,
		Phase:  protocol.PhaseMetadata,
		Result: valuePointer(result),
	}, nil
}

func resultObservation(
	ctx context.Context,
	backend *sqlite.Backend,
	contractID string,
	phase protocol.Phase,
	result protocol.Value,
	metrics *protocol.Value,
) (protocol.Observation, error) {
	state, err := readDatabaseState(ctx, backend)
	if err != nil {
		return protocol.Observation{}, err
	}
	return protocol.Observation{
		ID:      contractID,
		Status:  protocol.StatusObserved,
		Phase:   phase,
		Result:  valuePointer(result),
		DBState: valuePointer(state),
		Metrics: metrics,
	}, nil
}

func readDatabaseState(ctx context.Context, backend *sqlite.Backend) (protocol.Value, error) {
	articles, err := models.ArticleObjects.Using(backend).
		OrderBy(models.ArticleFields.ID.Asc()).
		All(ctx)
	if err != nil {
		return protocol.Value{}, fmt.Errorf("read database state: %w", err)
	}
	return databaseState(articles), nil
}

File: conformance/runners/django/scenarios.py
"""Explicit Django adapters for the initial GoDj compatibility contracts."""

from __future__ import annotations

from collections.abc import Callable, Iterator
from contextlib import contextmanager
from typing import Any

from django.conf import settings


_configured_by_godj = False


def configure_django() -> None:
    global _configured_by_godj

    if settings.configured:
        if _configured_by_godj:
            return
        raise RuntimeError(
            "refusing to use externally configured Django settings; "
            "the conformance runner requires its isolated in-memory default database"
        )

    settings.configure(
        DATABASES={
            "default": {
                "ENGINE": "django.db.backends.sqlite3",
                "NAME": ":memory:",
            }
        },
        DEFAULT_AUTO_FIELD="django.db.models.AutoField",
        INSTALLED_APPS=[
            "conformance.runners.django.migration_fixture.apps.GoDjMigrationFixtureConfig",
            "conformance.runners.django.migration_failure_fixture.apps.GoDjMigrationFailureFixtureConfig",
            "conformance.runners.django.migration_relation_fixture.apps.GoDjMigrationRelationFixtureConfig",
        ],
        LANGUAGE_CODE="en-us",
        SECRET_KEY="godj-conformance-not-a-secret",
        TIME_ZONE="UTC",
        USE_I18N=False,
        USE_TZ=True,
    )
    _configured_by_godj = True

    import django

    try:
        django.setup()
    except Exception:
        _configured_by_godj = False
        raise


configure_django()

from django.core.exceptions import FieldError  # noqa: E402
from django.db import connection, models  # noqa: E402

from .normalizer import PrimaryKey, normalize  # noqa: E402


class Article(models.Model):
    title = models.CharField(max_length=200)
    published = models.BooleanField(default=False)
    summary = models.CharField(max_length=200, null=True)

    class Meta:
        app_label = "godj_conformance"
        db_table = "godj_conformance_article"


FIXTURES = (
    Article(id=1, title="Alpine Guide", published=True, summary=None),
    Article(id=2, title="django Tips", published=False, summary="ORM"),
    Article(id=3, title="Django Deep Dive", published=True, summary=""),
    Article(id=4, title="Other", published=True, summary=None),
)


@contextmanager
def article_database() -> Iterator[None]:
    with connection.schema_editor() as editor:
        editor.create_model(Article)
    try:
        Article.objects.bulk_create(
            [
                Article(
                    id=row.id,
                    title=row.title,
                    published=row.published,
                    summary=row.summary,
                )
                for row in FIXTURES
            ]
        )
        yield
    finally:
        with connection.schema_editor() as editor:
            editor.delete_model(Article)


def _rows(queryset: Any) -> list[dict[str, Any]]:
    if not queryset.totally_ordered:
        raise AssertionError("ordered result contract requires total ordering")
    return [
        {
            "id": PrimaryKey(pk),
            "published": published,
            "summary": summary,
            "title": title,
        }
        for pk, title, published, summary in queryset.values_list(
            "id", "title", "published", "summary"
        )
    ]


def _database_state() -> dict[str, Any]:
    return {
        "articles": _rows(Article.objects.order_by("id")),
    }


def _observed(
    contract_id: str,
    result: Any,
    *,
    phase: str = "evaluation",
    metrics: Any | None = None,
) -> dict[str, Any]:
    return {
        "id": contract_id,
        "status": "observed",
        "phase": phase,
        "result": normalize(result),
        "error": None,
        "db_state": normalize(_database_state()),
        "metrics": normalize(metrics) if metrics is not None else None,
    }


def query_exact(contract_id: str) -> dict[str, Any]:
    with article_database():
        queryset = Article.objects.filter(title="Alpine Guide").order_by("id")
        return _observed(contract_id, _rows(queryset))


def query_ascii_icontains(contract_id: str) -> dict[str, Any]:
    with article_database():
        queryset = Article.objects.filter(title__icontains="django").order_by("id")
        return _observed(contract_id, _rows(queryset))


def query_chained_and(contract_id: str) -> dict[str, Any]:
    with article_database():
        queryset = (
            Article.objects.filter(title__icontains="django")
            .filter(published=True)
            .order_by("id")
        )
        return _observed(contract_id, _rows(queryset))


def query_chain_preserves_source(contract_id: str) -> dict[str, Any]:
    with article_database():
        base = Article.objects.filter(published=True)
        derived = base.filter(title="Django Deep Dive")
        result = {
            "base": _rows(base.order_by("id")),
            "derived": _rows(derived.order_by("id")),
        }
        return _observed(contract_id, result)


def query_order_and_limit(contract_id: str) -> dict[str, Any]:
    with article_database():
        queryset = Article.objects.order_by("-title", "id")[:2]
        return _observed(contract_id, _rows(queryset))


def query_empty_result(contract_id: str) -> dict[str, Any]:
    with article_database():
        queryset = Article.objects.filter(title="missing").order_by("id")
        return _observed(contract_id, _rows(queryset))


def query_isnull(contract_id: str) -> dict[str, Any]:
    with article_database():
        result = {
            "false": _rows(
                Article.objects.filter(summary__isnull=False).order_by("id")
            ),
            "true": _rows(
                Article.objects.filter(summary__isnull=True).order_by("id")
            ),
        }
        return _observed(contract_id, result)


def _field_error(
    contract_id: str,
    code: str,
    operation: Callable[[], Any],
) -> dict[str, Any]:
    with article_database():
        try:
            operation()
        except FieldError as exc:
            return {
                "id": contract_id,
                "status": "observed",
                "phase": "construction",
                "result": None,
                "error": {
                    "category": "field_error",
                    "code": code,
                    "python_type": f"{type(exc).__module__}.{type(exc).__qualname__}",
                    "message": str(exc),
                    "message_is_contract": False,
                },
                "db_state": normalize(_database_state()),
                "metrics": None,
            }
    raise AssertionError("expected django.core.exceptions.FieldError")


def query_unknown_field(contract_id: str) -> dict[str, Any]:
    return _field_error(
        contract_id,
        "unknown_field",
        lambda: Article.objects.filter(unknown_field="value"),
    )


def query_construction_has_no_io(contract_id: str) -> dict[str, Any]:
    with article_database():
        statements: list[str] = []

        def capture(execute, sql, params, many, context):
            statements.append(sql)
            return execute(sql, params, many, context)

        with connection.execute_wrapper(capture):
            queryset = Article.objects.filter(published=True).order_by("id")[:2]

        return _observed(
            contract_id,
            {"queryset_constructed": queryset is not None},
            phase="construction",
            metrics={"queries_during_construction": len(statements)},
        )


def query_unsupported_lookup(contract_id: str) -> dict[str, Any]:
    return _field_error(
        contract_id,
        "unsupported_lookup",
        lambda: Article.objects.filter(title__starts="Django"),
    )


def schema_model_metadata(contract_id: str) -> dict[str, Any]:
    fields = []
    for field in Article._meta.get_fields():
        fields.append(
            {
                "internal_type": field.get_internal_type(),
                "max_length": field.max_length,
                "name": field.name,
                "null": field.null,
                "primary_key": field.primary_key,
            }
        )
    return {
        "id": contract_id,
        "status": "observed",
        "phase": "metadata",
        "result": normalize(
            {
                "db_table": Article._meta.db_table,
                "fields": fields,
                "model_name": Article._meta.model_name,
            }
        ),
        "error": None,
        "db_state": None,
        "metrics": None,
    }


SCENARIOS: dict[str, Callable[[str], dict[str, Any]]] = {
    "django.query.exact": query_exact,
    "django.query.ascii_icontains": query_ascii_icontains,
    "django.query.chained_and": query_chained_and,
    "django.query.chain_preserves_source": query_chain_preserves_source,
    "django.query.order_and_limit": query_order_and_limit,
    "django.query.empty_result": query_empty_result,
    "django.query.isnull": query_isnull,
    "django.query.unknown_field": query_unknown_field,
    "django.query.construction_has_no_io": query_construction_has_no_io,
    "django.query.unsupported_lookup": query_unsupported_lookup,
    "django.schema.model_metadata": schema_model_metadata,
}
